homeassistant/components/modbus/__init__.py

```diff
@@ -88,163 +97,219 @@
+    _LOGGER.debug("registering hubs")
+    for client_config in config[MODBUS_DOMAIN]:
+        hub_collect[client_config[CONF_NAME]] = ModbusHub(client_config, hass.loop)
 
     def stop_modbus(event):
         """Stop Modbus service."""
         for client in hub_collect.values():
-            client.close()
+            del client
 
     def start_modbus(event):
         """Start Modbus service."""
         for client in hub_collect.values():
-            client.connect()
+            _LOGGER.debug("setup hub %s", client.name)
+            client.setup()
 
         hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, stop_modbus)
 
         # Register services for modbus
-        hass.services.register(
-            DOMAIN,
+        hass.services.async_register(
+            MODBUS_DOMAIN,
             SERVICE_WRITE_REGISTER,
             write_register,
             schema=SERVICE_WRITE_REGISTER_SCHEMA,
         )
-        hass.services.register(
-            DOMAIN, SERVICE_WRITE_COIL, write_coil, schema=SERVICE_WRITE_COIL_SCHEMA
+        hass.services.async_register(
+            MODBUS_DOMAIN,
+            SERVICE_WRITE_COIL,
+            write_coil,
+            schema=SERVICE_WRITE_COIL_SCHEMA,
         )
 
-    def write_register(service):
+    async def write_register(service):
         """Write Modbus registers."""
         unit = int(float(service.data[ATTR_UNIT]))
         address = int(float(service.data[ATTR_ADDRESS]))
         value = service.data[ATTR_VALUE]
         client_name = service.data[ATTR_HUB]
         if isinstance(value, list):
-            hub_collect[client_name].write_registers(
+            await hub_collect[client_name].write_registers(
                 unit, address, [int(float(i)) for i in value]
             )
         else:
-            hub_collect[client_name].write_register(unit, address, int(float(value)))
+            await hub_collect[client_name].write_register(
+                unit, address, int(float(value))
+            )
 
-    def write_coil(service):
+    async def write_coil(service):
         """Write Modbus coil."""
         unit = service.data[ATTR_UNIT]
         address = service.data[ATTR_ADDRESS]
         state = service.data[ATTR_STATE]
         client_name = service.data[ATTR_HUB]
-        hub_collect[client_name].write_coil(unit, address, state)
+        await hub_collect[client_name].write_coil(unit, address, state)
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, start_modbus)
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, start_modbus)
 
     return True
 
 
 class ModbusHub:
     """Thread safe wrapper class for pymodbus."""
 
-    def __init__(self, modbus_client, name):
+    def __init__(self, client_config, main_loop):
         """Initialize the Modbus hub."""
-        self._client = modbus_client
-        self._lock = threading.Lock()
-        self._name = name
+        _LOGGER.debug("Preparing setup: %s", client_config)
```

Comment:
We don't want side effects, eg logging, in init method. Please move it outside, or remove it.